File: windows-app/Plantoir.Core/Scripting/TaskMilestones.cs
namespace Plantoir.Core.Scripting;

/// <summary>A label shown to the teacher and the marker searched for in output.</summary>
public sealed record TaskMilestone(string Label, string Marker);

/// <summary>
/// The ordered milestone lists for each task, with markers matched against
/// what the WINDOWS launchers really print (the .ps1 output differs from the
/// .sh output in its host-side lines; the in-container Python lines are
/// shared). Every label ends with an ellipsis — the shared invariant.
/// </summary>
public static class TaskMilestones
{
    public static readonly IReadOnlyList<TaskMilestone> CourseCreation = new[]
    {
        new TaskMilestone("Getting things ready…", "Detected host timezone offset"),
        new TaskMilestone("Getting ready…", "Welcome to the Course Setup Script"),
        new TaskMilestone("Preparing your course folder…", "'Media' folder"),
        new TaskMilestone("Choosing appearance…", "Quartz Locale"),
        new TaskMilestone("Setting up sections…", "timetable section numbers"),
        new TaskMilestone("Creating folders and files…", "Select folders/files to HIDE"),
        new TaskMilestone("Finishing up…", "set up successfully"),
    };

    public static readonly IReadOnlyList<TaskMilestone> ExampleCourse = new[]
    {
        // A THIRD marker the mac's example-course list does not have, and it
        // passes MilestoneContractTests.EveryTaskShowsTheSharedStepsTheMacShows
        // by a margin nobody designed. That test compares only the markers
        // classified in app-rules.json → markerOrigins.origins, and the key
        // there is "Host timezone offset" — not this string — so this one
        // filters out of BOTH sides before the comparison and the two lists
        // agree about what is left.
        //
        // Which means: if anyone ever "tidies" that origins key to match the
        // full text a launcher prints, this test starts failing, and the cause
        // will look nothing like the change that caused it. The fix would then
        // be a decision about whether the two example-course lists should show
        // the same steps — not a change to the readout. Flagged from the mac in
        // GitHub issue #121 and written down here, because a closed issue is
        // not where somebody editing a marker string will look.
        new TaskMilestone("Getting things ready…", "Detected host timezone offset"),
        new TaskMilestone("Copying the example course…", "Example Course installed to"),
        new TaskMilestone("Finishing up…", "EXAMPLE_COURSE_CODE="),
    };

    public static readonly IReadOnlyList<TaskMilestone> Preview = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Applying your settings…", "Updated pageTitle"),
        new TaskMilestone("Preparing components…", "Installing dependencies"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        // "Launching Quartz preview" — build_site.py's own print — fires BEFORE
        // `quartz build --serve` even starts, so it used to complete every
        // milestone at once and pin the bar on this step for the whole real
        // build (TODO.md item 1, found 2026-08-19). "Done processing" is
        // patches/build.ts's own line, printed once the fresh site is actually
        // written to disk — verified against a real `preview.ps1 --build-only`
        // transcript 2026-08-23 ("Quartz v4.5.0" then "Done processing N files
        // in Ns"). markerOrigins.origins in contracts/app-rules.json calls
        // this "elsewhere" (Quartz's own output, not shared Python).
        new TaskMilestone("Opening the preview…", "Done processing"),
    };

    public static readonly IReadOnlyList<TaskMilestone> Deploy = new[]
    {
        new TaskMilestone("Getting things ready…", "Host timezone offset"),
        new TaskMilestone("Starting up…", "from this PC"),
        new TaskMilestone("Checking your site…", "Deploying from local build"),
        new TaskMilestone("Connecting to Netlify…", "Netlify site"),
        new TaskMilestone("Comparing what changed…", "delta deploy manifest"),
        new TaskMilestone("Uploading your pages…", "Delta deploy created"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    /// <summary>Rebuild + publish presented as ONE task with one bar.</summary>
    public static readonly IReadOnlyList<TaskMilestone> BuildAndDeploy = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        new TaskMilestone("Connecting to Netlify…", "Netlify site"),
        new TaskMilestone("Uploading your pages…", "Delta deploy created"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    /// <summary>
    /// Folder-mode publish: the site is already built on the host, so the
    /// launcher only mirrors changed files into the chosen folder — quick,
    /// containerless, and never mentioning Netlify (row 102d).
    /// </summary>
    public static readonly IReadOnlyList<TaskMilestone> DeployToFolder = new[]
    {
        new TaskMilestone("Checking your site…", "Host timezone offset"),
        new TaskMilestone("Copying your files…", "to a folder"),
        new TaskMilestone("Finishing up…", "PUBLISHED_FOLDER="),
    };

    /// <summary>Rebuild + folder publish presented as ONE task with one bar.</summary>
    public static readonly IReadOnlyList<TaskMilestone> BuildAndDeployToFolder = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        new TaskMilestone("Copying your files…", "to a folder"),
        new TaskMilestone("Finishing up…", "PUBLISHED_FOLDER="),
    };

    /// <summary>
    /// Cloudflare publish. The phrases match what deploy.py prints on that
    /// path — and, like folder mode, never say Netlify.
    /// </summary>
    public static readonly IReadOnlyList<TaskMilestone> DeployToCloudflare = new[]
    {
        new TaskMilestone("Getting things ready…", "Host timezone offset"),
        new TaskMilestone("Starting up…", "from this PC"),
        new TaskMilestone("Checking your site…", "Deploying from local build"),
        new TaskMilestone("Connecting to Cloudflare…", "Cloudflare project"),
        new TaskMilestone("Uploading your pages…", "Uploading the built site"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    /// <summary>Rebuild + Cloudflare publish presented as ONE task with one bar.</summary>
    public static readonly IReadOnlyList<TaskMilestone> BuildAndDeployToCloudflare = new[]
    {
        new TaskMilestone("Getting things ready…", "Running the website builder on this PC"),
        new TaskMilestone("Gathering your content…", "Copying shared folders"),
        new TaskMilestone("Building your site…", "Quartz v4"),
        new TaskMilestone("Connecting to Cloudflare…", "Cloudflare project"),
        new TaskMilestone("Uploading your pages…", "Uploading the built site"),
        new TaskMilestone("Finishing up…", "Deploy complete"),
    };

    public static IEnumerable<IReadOnlyList<TaskMilestone>> AllLists
    {
        get
        {
            yield return CourseCreation;
            yield return ExampleCourse;
            yield return Preview;
            yield return Deploy;
            yield return BuildAndDeploy;
            yield return DeployToFolder;
            yield return BuildAndDeployToFolder;
            yield return DeployToCloudflare;
            yield return BuildAndDeployToCloudflare;
        }
    }
}

File: mac-app/QuartzTeachers/Scripting/TaskMilestones.swift
import Foundation

/// One step of a long-running task: the text shown to the teacher, and
/// the marker in the output that means the step has been reached.
struct TaskMilestone {

    // MARK: - Stored properties

    /// Brief, meaningful label, e.g. "Building your site".
    let label: String

    /// A phrase that appears in the output when this step begins.
    let marker: String
}

/// The ordered milestones for each kind of task. Progress is "how many
/// milestones have been reached", so the bar advances in real steps
/// rather than spinning indefinitely.
enum TaskMilestones {

    // MARK: - Stored properties

    /// Creating a course: the wizard's own progression.
    static let courseCreation: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Getting ready…", marker: "Welcome to the Course Setup Script"),
        TaskMilestone(label: "Preparing your course folder…", marker: "'Media' folder"),
        TaskMilestone(label: "Choosing appearance…", marker: "Quartz Locale"),
        TaskMilestone(label: "Setting up sections…", marker: "timetable section numbers"),
        TaskMilestone(label: "Creating folders and files…", marker: "Select folders/files to HIDE"),
        TaskMilestone(label: "Finishing up…", marker: "set up successfully"),
    ]

    /// Installing the example course.
    static let exampleCourse: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Getting things ready…", marker: "Starting container if needed"),
        TaskMilestone(label: "Copying the example course…", marker: "Example Course installed to"),
        TaskMilestone(label: "Finishing up…", marker: "EXAMPLE_COURSE_CODE="),
    ]

    /// Previewing a section.
    static let preview: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Starting container if needed"),
        TaskMilestone(label: "Gathering your content…", marker: "Copying shared folders"),
        TaskMilestone(label: "Applying your settings…", marker: "Updated pageTitle"),
        TaskMilestone(label: "Preparing components…", marker: "Installing dependencies"),
        TaskMilestone(label: "Building your site…", marker: "Quartz v4"),
        // "Launching Quartz preview" — build_site.py's own print — fires BEFORE
        // `quartz build --serve` even starts, so it used to complete every
        // milestone at once and pin the bar on this step for the whole real
        // build (found 2026-08-19). "Done processing" is patches/build.ts's own
        // line, printed once the fresh site is actually written to disk —
        // verified against a real `preview.ps1 --build-only` transcript on
        // Windows 2026-08-23 ("Quartz v4.5.0" then "Done processing N files in
        // Ns"). Authored on Windows; built, tested and regenerated into
        // app-rules.json on the mac 2026-09-01 (13da5319).
        TaskMilestone(label: "Opening the preview…", marker: "Done processing"),
    ]

    /// Publishing a section to Netlify.
    static let deploy: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Ensuring container is running"),
        TaskMilestone(label: "Checking your site…", marker: "Deploying from local build"),
        TaskMilestone(label: "Connecting to Netlify…", marker: "Netlify site"),
        TaskMilestone(label: "Comparing what changed…", marker: "delta deploy manifest"),
        TaskMilestone(label: "Uploading your pages…", marker: "Delta deploy created"),
        TaskMilestone(label: "Finishing up…", marker: "Deploy complete"),
    ]

    /// Publishing when the site has to be rebuilt first — one task from
    /// the teacher's point of view, so one progress bar.
    static let buildAndDeploy: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Starting container if needed"),
        TaskMilestone(label: "Gathering your content…", marker: "Copying shared folders"),
        TaskMilestone(label: "Building your site…", marker: "Quartz v4"),
        TaskMilestone(label: "Connecting to Netlify…", marker: "Netlify site"),
        TaskMilestone(label: "Uploading your pages…", marker: "Delta deploy created"),
        TaskMilestone(label: "Finishing up…", marker: "Deploy complete"),
    ]

    /// Deploying a section to Cloudflare Pages. Netlify is never involved,
    /// so the progress must not mention it. There is no "comparing what
    /// changed" step: Cloudflare's own tool works out which assets it
    /// already holds, rather than the manifest exchange Netlify uses.
    static let deployToCloudflare: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Ensuring container is running"),
        TaskMilestone(label: "Checking your site…", marker: "Deploying from local build"),
        TaskMilestone(label: "Connecting to Cloudflare…", marker: "Cloudflare project"),
        TaskMilestone(label: "Uploading your pages…", marker: "Uploading the built site"),
        TaskMilestone(label: "Finishing up…", marker: "Deploy complete"),
    ]

    /// Deploying to Cloudflare when the site has to be rebuilt first.
    static let buildAndDeployToCloudflare: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Starting container if needed"),
        TaskMilestone(label: "Gathering your content…", marker: "Copying shared folders"),
        TaskMilestone(label: "Building your site…", marker: "Quartz v4"),
        TaskMilestone(label: "Connecting to Cloudflare…", marker: "Cloudflare project"),
        TaskMilestone(label: "Uploading your pages…", marker: "Uploading the built site"),
        TaskMilestone(label: "Finishing up…", marker: "Deploy complete"),
    ]

    /// Publishing a section to a folder on this Mac. Netlify is never
    /// involved, and neither is the container: the built site already
    /// sits on this Mac, so the whole publish is a quick local copy.
    static let deployToFolder: [TaskMilestone] = [
        TaskMilestone(label: "Checking your site…", marker: "Host timezone offset"),
        TaskMilestone(label: "Copying your files…", marker: "to a folder"),
        TaskMilestone(label: "Finishing up…", marker: "PUBLISHED_FOLDER="),
    ]

    /// Publishing to a folder when the site has to be rebuilt first.
    static let buildAndDeployToFolder: [TaskMilestone] = [
        TaskMilestone(label: "Getting this Mac ready…", marker: "Setting up this Mac"),
        TaskMilestone(label: "Building your website builder…", marker: "Building your website builder"),
        TaskMilestone(label: "Starting up…", marker: "Starting container if needed"),
        TaskMilestone(label: "Gathering your content…", marker: "Copying shared folders"),
        TaskMilestone(label: "Building your site…", marker: "Quartz v4"),
        TaskMilestone(label: "Copying your files…", marker: "to a folder"),
        TaskMilestone(label: "Finishing up…", marker: "PUBLISHED_FOLDER="),
    ]

    /// Every list above, under the name the contract writes it as.
    ///
    /// `AppRulesContract` builds its readout from THIS, rather than from a
    /// second copy of the list of lists, and that is the whole point of it
    /// existing. Until 2026-09-08 that contract kept a private array naming
    /// eight while this enum had nine: `exampleCourse` reached no readout, so
    /// the two shared-python markers only it carries — "Example Course
    /// installed to" and "EXAMPLE_COURSE_CODE=" — were classified by nobody,
    /// and the test that walks the readout could not see that they were
    /// missing. A readout assembled from a hand-kept list of what to read is a
    /// readout that can silently omit things.
    ///
    /// **The hole this does not close**, said plainly rather than left to be
    /// discovered: a new `static let` added above and not added here is still
    /// invisible. Swift cannot enumerate an enum's static properties at
    /// runtime, so nothing can catch that automatically — Windows' `AllLists`
    /// has the identical gap. Adding a list means adding it in both places.
    static let allLists: [(name: String, milestones: [TaskMilestone])] = [
        ("courseCreation", courseCreation),
        ("exampleCourse", exampleCourse),
        ("preview", preview),
        ("deploy", deploy),
        ("buildAndDeploy", buildAndDeploy),
        ("deployToCloudflare", deployToCloudflare),
        ("buildAndDeployToCloudflare", buildAndDeployToCloudflare),
        ("deployToFolder", deployToFolder),
        ("buildAndDeployToFolder", buildAndDeployToFolder),
    ]

}
